File: 프로그래머스/3/12927. 야근 지수/야근 지수.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

long long solution(int n, vector<int> works) {
    long long total = 0;
    
    for (int w : works) {
        total += w;
    }
    
    // 일을 다 처리할 수 있으면 피로도는 0
    if (total <= n) return 0;
    
    priority_queue<int> pq;
    
    for (int w : works) {
        pq.push(w);
    }
    
    while (n--) {
        int top = pq.top();
        pq.pop();
        pq.push(top - 1);
    }
    
    long long answer = 0;
    
    while (!pq.empty()) {
        long long x = pq.top();
        pq.pop();
        answer += x * x;
    }
    
    return answer;
}

File: 프로그래머스/3/12927. 야근 지수/야근 지수.java
import java.util.Collections;
import java.util.PriorityQueue;

class Solution {
    public long solution(int n, int[] works) {
        long total = 0;
        
        for (int work : works) {
            total += work;
        }
        
        // 남은 시간 안에 일을 전부 끝낼 수 있으면 야근 피로도는 0
        if (total <= n) return 0;
        
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        
        for (int work : works) {
            pq.offer(work);
        }
        
        while (n-- > 0) {
            int max = pq.poll();
            pq.offer(max - 1);
        }
        
        long answer = 0;
        
        while (!pq.isEmpty()) {
            long x = pq.poll();
            answer += x * x;
        }
        
        return answer;
    }
}